src/ZVal.h

```diff
@@ -0,0 +1,505 @@
+// See the file "COPYING" in the main distribution directory for copyright.
+
+// Values used in ZAM execution, and also for representing records and
+// vectors during interpreter execution.
+
+#pragma once
+
+#include <unordered_set>
+
+#include "Dict.h"
+#include "Expr.h"
+#include "IntrusivePtr.h"
+
+ZEEK_FORWARD_DECLARE_NAMESPACED(StringVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(AddrVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(SubNetVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(File, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(Func, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(ListVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(OpaqueVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(PatternVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(TableVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(RecordVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(VectorVal, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(Type, zeek);
+ZEEK_FORWARD_DECLARE_NAMESPACED(Val, zeek);
+
+namespace zeek {
+
+// Only needed for compiled code.
+class IterInfo;
+
+typedef std::vector<zeek::ValPtr> val_vec;
+
+// A bit of this mirrors BroValUnion, but BroValUnion captures low-level
+// representation whereas we aim to keep Val structure for more complex
+// Val's.
+//
+// Ideally we'd use IntrusivePtr's for memory management, but we can't
+// given we have a union and thus on destruction C++ doesn't know which
+// member flavor to destruct.
+//
+// Note that a ZAMValUnion by itself is ambiguous: it doesn't track its
+// type.  This makes them consume less memory and cheaper to copy.  It
+// does however require a separate way to determine the type.  Generally
+// this is doable using surrounding context, or can be statically determined
+// in the case of optimization/compilation.
+//
+// An alternative would be to use std::variant, but (1) it tracks the
+// variant type, and (2) it won't allow access to the managed_val member,
+// which not only simplifies memory management but also is required for
+// sharing of ZAM frame slots.
+union ZAMValUnion {
```

Comment:
I do think we should switch this to a `std::variant` to make it type-safe and all memory-managed. Would we then actually still need the notion of being "managed"?

Could we do (2)  through a dispatcher method a returning an object pointer?

Reply:
I resist this.  It means that every low-level copy of a ZAMValUnion is more expensive, and they cost twice as much to store.  Plus sharing of ZAM frame slots provides a significant performance gain: some fully inlined event handlers get a 10x savings in frame size (which is also a speed savings due to initialization issues). I think the uneasiness of non-type-safety is worth it in this case for the performance benefits

Reply:
I'm quite reluctant to merge new code that's neither type-safe nor automatically memory managed. It's not only "uneasiness", but also a major maintenance burden and a barrier to entry for new people. We've been working very hard to get away from that, and I want to be careful that we don't end up moving the wrong way.

Of course performance is paramount for an optimizer, so that's a perfectly valid argument. However, I don't know if it's as bad as you suggest. Honestly, I've given up on making these kinds of assumptions on performance without measuring, I've been wrong often enough about trying to guess how something will turn out. For example, expensive copies may be avoidable by moving instances; storing a heap-allocated object twice the size might not matter because of malloc alignment; and storing/accessing more information can be largely invisible if it all fits into a cache line.

In short: I don't know the answer yet either, but I wouldn't just rule out yet that a safer way do it might be feasible, too.

Reply:
Let me try a different tack: Are we sure we need the full ZAMValUnion at the interpreter level (vs moving it down into the optimizer-only code)? I know we need to have a version of `Z_vector` / `Z_record` inside the `BroValUnion` and I know those depend on BroValUnion in turn. But could there be a different way of structuring this so we extend `BroValUnion` without pulling in the full `BroValUnion`?

Reply:
I've been noodling on this but I'm not seeing how to make such a path work.  There are multiple challenges:

1. For the foreseeable future, we'll want compiled code and interpreted code able to live together (each can call the other).  This requires that they share underlying data types, since otherwise the types need to be inter-converted, which is not only inefficient but more fundamentally leads to broken semantics (aggregates no longer being shallow-copied).
2. There are major performance gains in representing Zeek vectors and records as `std::vector`'s of some sort of low-level type, rather than `Val*`'s.  Doing so enables avoiding a bunch of `Val` construction/destruction.
3. To preserve current APIs (fundamental for the script interpreter), `ZAM_vector` and `ZAM_record` need to be able to return `Val*`'s for their elements/fields. While `ZAMValUnion` supports doing so, `BroValUnion` does not, and it lacks the hooks for adding this in a manner that preserves the shallow-copy semantics of aggregates.  We could get those hooks by making all of the low-level types in `BroValUnion` derived from `Obj`, but per my comment above unfortunately I think this is a heavy lift :-( .